torch/distributed/fsdp/flat_param.py

```diff
@@ -847,6 +847,37 @@ def _free_low_precision_sharded_param(self):
         self._check_low_precision_shard()
         _free_storage(self.flat_param._mp_shard)  # type: ignore[attr-defined]
 
+    @torch.no_grad()
+    def unshard_grad(self):
+        if not self.uses_sharded_strategy:
+            self._use_unsharded_grad_views()
+            return
+        flat_param = self.flat_param
+        self._check_unsharded(flat_param)
+        padded_unsharded_grad = torch.empty(
+            flat_param._padded_unsharded_size,  # type: ignore[attr-defined]
+            device=self.device,
+        )
+        if flat_param.grad is None:
+            flat_param._saved_grad_shard = None  # type: ignore[attr-defined]
+            sharded_grad = torch.zeros_like(flat_param)  # type: ignore[attr-defined]
+        else:
+            self._check_sharded(flat_param.grad)
+            flat_param._saved_grad_shard = flat_param.grad  # type: ignore[attr-defined]
```

Comment:
curious why we intentionally fill "flat_param._saved_grad_shard" here?

Reply:
We already use `_saved_grad_shard` for saving the sharded gradient during the backward pass. Here, I use the same variable, now for a different purpose: saving the sharded gradient during `summon_full_params(with_grads=True)`. This is just to avoid creating an entirely new variable also for saving the sharded gradient.